packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,280 @@
+        lastProcessedAuctionId = lastProcessedAuctionId_;
+        nextProposalRewardFirstAuctionId = nextProposalRewardFirstAuctionId_;
+        params = rewardParams;
+    }
+
+    struct Temp {
+        uint256 numEligibleVotes;
+        uint256 rewardPerProposal;
+        uint256 rewardPerVote;
+    }
+
+    function setParams(RewardParams calldata newParams) public onlyOwner {
+        params = newParams;
+    }
+
+    function getParams() public view returns (RewardParams memory) {
+        return params;
+    }
+
+    function updateRewardsForAuctions(uint256 lastNounId) public {
+        uint256 currentlyAuctionedNounId = auctionHouse.auction().nounId;
+        require(lastNounId < currentlyAuctionedNounId, 'lastNounId must be settled');
+
+        uint256 lastProcessedAuctionId_ = lastProcessedAuctionId;
+        require(lastNounId > lastProcessedAuctionId_, 'lastNounId must be higher');
+        lastProcessedAuctionId = lastNounId;
+
+        INounsAuctionHouseV2.Settlement memory settlement;
+        for (uint256 nounId = lastProcessedAuctionId_ + 1; nounId <= lastNounId; nounId++) {
+            // TODO maybe can be optimized to one call with large range
+            settlement = auctionHouse.getSettlements(nounId, nounId + 1, false)[0];
+            if (settlement.clientId > 0) {
+                clientBalances[settlement.clientId] += (settlement.amount * params.auctionRewardBps) / 10_000;
+            }
+        }
+    }
+
+    /**
+     * @param lastProposalId id of the last proposal to include in the rewards distribution. all proposals up to and
+     * including this id must have ended voting.
+     * @param lastAuctionedNounId the noun id that was auctioned when proposal with id `lastProposalId` was created.
+     * The auction revenue up to this noun id are included in the bounty.
+     * @param votingClientIds array of client ids that were used to vote on of all eligible the eligible proposals in
+     * this rewards distribution
+     */
+    function updateRewardsForProposalWritingAndVoting(
+        uint32 lastProposalId,
+        uint256 lastAuctionedNounId,
+        uint32[] calldata votingClientIds
+    ) public {
+        require(lastProposalId <= nounsDAO.proposalCount(), 'bad lastProposalId');
+        require(lastProposalId >= nextProposalIdToReward, 'bad lastProposalId');
+
+        NounsDAOV3Types.ProposalForRewards[] memory proposals = nounsDAO.proposalDataForRewards(
+            nextProposalIdToReward,
+            lastProposalId,
+            votingClientIds
+        );
+        nextProposalIdToReward = lastProposalId + 1;
+
+        NounsDAOV3Types.ProposalForRewards memory lastProposal = proposals[proposals.length - 1];
+
+        uint256 auctionRevenue = getAuctionRevenueBetweenNouns({
+            firstNounId: nextProposalRewardFirstAuctionId,
+            oneAfterLastNounId: lastAuctionedNounId,
+            endTimestamp: lastProposal.creationTimestamp
```

Comment:
feels a bit awkward that `lastProposal` could be a proposal where creationTimestamp is zero. I believe the code will revert in `getAuctionRevenueBetweenNouns` because of the first require statement there.

maybe we should have a more explicit requirement that lastProposal has to have a non zero creation timestamp, wdyt?

Reply:
maybe, though for it to happen we would need to initially set `nextProposalRewardFirstAuctionId` to a proposal before creation timestamp is being set. which we won't do.
still think it's helpful?

Reply:
let's clarify further what you have in mind; are you saying we will first deploy the DAO upgrade, then wait for the first proposal with the timestamp to be posted, and only then we will deploy the rewards contract?

Reply:
@eladmallel yes, we'll first deploy the DAO upgrade & Noracle.
we need both proposals with timestamps and auctions with settlements.
once those are done, we deploy the rewards contract and and put a proposal to fund it either immediately or once it starts to accumulate some balances